src/utils.js

```diff
@@ -66,9 +76,342 @@ export function bytesAreEqual(v1, v2) {
   if (v1.length !== v2.length) {
     return false;
   }
-  let mismatch = false;
+  let mismatch = 0;
   for (let i = 0; i < v1.length; i++) {
-    mismatch &= v1[i] !== v2[i];
+    mismatch |= v1[i] ^ v2[i];
   }
   return ! mismatch;
 }
+
+// The `BufferReader` and `BufferWriter` classes are helpers for dealing with the
+// binary struct format that's used for various TLS message.  Think of them as a
+// buffer with a pointer to the "current position" and a bunch of helper methods
+// to read/write structured data and advance said pointer.
+//
+// XXX TODO: these should throw a "decode error" alert when reading past the end
+// of the buffer, as it indicates a malformed TLS message of some sort.
+
+class BufferWithPointer {
+  constructor(buf) {
+    assertIsBytes(buf);
+    this._buffer = buf;
+    this._dataview = new DataView(buf.buffer, buf.byteOffset, buf.byteLength);
+    this._pos = 0;
+  }
+
+  resize(size) {
+    assert(size > this.length(), 'cant resize BufferWithPointer to be smaller');
+    const newbuf = new Uint8Array(size);
+    newbuf.set(this._buffer.slice(0, this.tell()), 0);
+    this._buffer = newbuf;
+    this._dataview = new DataView(newbuf.buffer, newbuf.byteOffset, newbuf.byteLength);
+  }
+
+  length() {
+    return this._buffer.byteLength;
+  }
+
+  tell() {
+    return this._pos;
+  }
+
+  seek(pos) {
+    this._pos = pos;
+    assert(this._pos <= this.length(), 'do not seek past end of buffer');
+  }
+
+  incr(offset) {
+    this.seek(this._pos + offset);
+  }
+
+  slice(start = 0, end = this.tell()) {
+    return this._buffer.slice(start, end);
+  }
+}
+
+
+export class BufferReader extends BufferWithPointer {
+
+  hasMoreBytes() {
+    return this.tell() < this.length();
+  }
+
+  readBytes(length) {
+    const slice = this.slice(this.tell(), this.tell() + length);
+    this.incr(length);
+    return slice;
+  }
+
+  readRemainingBytes() {
```

Comment:
In TLS, this is probably a hazard.  You should always know how many bytes you are reading.

Reply:
👍 ; from a quick grep it looks like this is actually a leftover that is no longer used, so I guess I refactored away whatever hazard was using it...